compiler/runtime/OMRRSSReport.cpp

```diff
@@ -0,0 +1,354 @@
+ * License, version 2 with the OpenJDK Assembly Exception [2].
+ *
+ * [1] https://www.gnu.org/software/classpath/license.html
+ * [2] https://openjdk.org/legal/assembly-exception.html
+ *
+ * SPDX-License-Identifier: EPL-2.0 OR Apache-2.0 OR GPL-2.0-only WITH Classpath-exception-2.0 OR GPL-2.0-only WITH OpenJDK-assembly-exception-1.0
+ *******************************************************************************/
+
+#include "runtime/CodeCacheManager.hpp"
+#include "runtime/CodeCache.hpp"
+
+OMR::RSSReport *OMR::RSSReport::_instance = NULL;
+
+const char *OMR::RSSItem::itemNames [] =
+      {
+      "Unknown",
+      "HEADER",
+      "ALIGNMENT",
+      "COLD BLOCKS",
+      "OVERESTIMATE",
+      "OOL",
+      "SNIPPET"
+      };
+
+void
+OMR::RSSRegion::addRSSItem(OMR::RSSItem *item, int32_t threadId, const char* methodName)
+   {
+#ifdef DEBUG_RSS_REPORT
+   TR_VerboseLog::writeLineLocked(TR_Vlog_PERF, "RSS adding item addr=%p size=%zu type=%s thread=%d method=%s", item->_addr, item->_size,
+                                  OMR::RSSItem::itemNames[item->_type], threadId,
+                                  methodName ? methodName : "no method");
+#endif
+
+   uint8_t *address = item->_addr;
+   TR_PersistentList<TR::DebugCounterAggregation> counters = item->_counters;
+
+   TR_ASSERT_FATAL(address, "Address should not be null");
+   TR_ASSERT_FATAL(_pageSize >=0, "Page size should be set");
+
+   size_t addressPage = (uintptr_t)address / _pageSize;
+   size_t startPage = (uintptr_t)_start / _pageSize;
+   int32_t offset = static_cast<int32_t>((_grows == lowToHigh) ? addressPage - startPage : startPage - addressPage);
+
+   TR_ASSERT_FATAL(offset >= 0, "Offset should be >= 0\n");
+
+   addToListSorted(&_pageMap[offset], item);
+
+   // split across pages if necessary
+   size_t tillPageEnd = _pageSize - ((uintptr_t)address % _pageSize);
+
+   OMR::RSSItem *newItem = item;
+
+   while (newItem->_size > tillPageEnd)
+      {
+      size_t newSize = newItem->_size - tillPageEnd;
+      newItem->_size = tillPageEnd;
+      address += newItem->_size;
+
+      newItem = new (TR::Compiler->persistentMemory()) OMR::RSSItem(item->_type, address, newSize, counters);
+      offset = (_grows == lowToHigh) ? ++offset : --offset;
+
+      TR_ASSERT_FATAL(offset >= 0, "Got negative offset %d for addr=%p size=%zu type=%s\n", offset,
+                      item->_addr, item->_size, OMR::RSSItem::itemNames[item->_type]);
+
+#ifdef DEBUG_RSS_REPORT
+      TR_VerboseLog::writeLineLocked(TR_Vlog_PERF, "RSS adding overflow item addr=%p size=%zu type=%s thread=%d method=%s",
+                                     newItem->_addr, newItem->_size,
+                                     OMR::RSSItem::itemNames[newItem->_type], threadId,
+                                     methodName ? methodName : "no method");
+#endif
+      addToListSorted(&_pageMap[offset], newItem);
+       tillPageEnd = _pageSize;
+      }
```

Comment:
Actually, I think this code produces least number of checks: we add original item, then check if it was too big, then correct it's size and create/add a new one. Maybe I'll just add a comment clarifying that?

Reply:
Please note that the first(original) item  is different from all subsequent items. I think the code above is the most concise and avoids separate checking and size calculations for the first item.

Reply:
Given that the RSSItems are always page sized (except for possibly the last (including the only) one), although something like this:
```
   size_t remainingSize = item->_size;
   size_t tillPageEnd = _pageSize - ((uintptr_t)address % _pageSize);
   
   if (remainingSize > _pageSize)
      item->_size = _pageSize;
   
   remainingSize -= item->_size
      
   addToListSorted(&_pageMap[offset], item);
   
   while (remainingSize)
      {
      size_t newSize = (remainingSize > _pageSize) ? _pageSize : tillPageEnd;
      
      address += newSize;
      OMR::RSSItem *newItem = new (TR::Compiler->persistentMemory()) OMR::RSSItem(item->_type, address, newSize, counters);
      offset = (_grows == lowToHigh) ? ++offset : --offset;
      
      TR_ASSERT_FATAL(offset >= 0, "Got negative offset %d for addr=%p size=%zu type=%s\n", offset,
                         item->_addr, item->_size, OMR::RSSItem::itemNames[item->_type]);
      
   #ifdef DEBUG_RSS_REPORT
         TR_VerboseLog::writeLineLocked(TR_Vlog_PERF, "RSS adding overflow item addr=%p size=%zu type=%s thread=%d method=%s",
                                        newItem->_addr, newItem->_size,
                                        OMR::RSSItem::itemNames[newItem->_type], threadId,
                                        methodName ? methodName : "no method");
   #endif
   
      addToListSorted(&_pageMap[offset], newItem);
      
      remainingSize -= newSize;
      }
```
incurs something like one additional comparison operation + one additional subtraction operation per iteration, it is clear what the size of the items are right when they are constructed, and once inserted into the list it isn't mutated. `newItem` is also always a new item.

Essentially, unless this code path is shown to be an extremely hot path that needs every drop of performance (which is unlikely given that this is for reporting and analysis), I don't think the post-insertion size fixup is worth the readability cost.

Reply:
Sure, I wanted to avoid those extra checks in the code. But perhaps your code is more general in case we change List implementation and start adding items vs pointers to them. I will rewrite.

Reply:
The above loop is not correct. For example, `address += newSize; `before creating the new item would be wrong since `address` would point to the end of the new item and not the beginning. To be honest, I think the original loop was better. There is nothing wrong with modifying previously inserted item if we know that we inserted a pointer to it. Just couple of comments would make the approach clear: we insert an item and as long as it does not fit into a page we continue creating new items while modifying previous one.

Reply:
Improved the old version to make it more clear and also add a separate commit with a more sequential version. So we can choose now :)

Reply:
Ah I see, I didn't realize that the first item can also be less than a page size (unless it was page aligned to begin with).

I mean my big issue was the shallow copy of the counters, which has since been fixed, so with the added comments in commit `f30ce76fd6` I'm less opposed to the original approach; at the very least it's much easier to understand now.